crates/factors/factors/src/factor_source.rs

```diff
@@ -155,6 +157,61 @@ impl FactorSource {
     }
 }
 
+impl FactorSource {
+    pub fn with_details(
+        factor_source_kind: FactorSourceKind,
+        mnemonic_with_passphrase: MnemonicWithPassphrase,
+        name: String,
+        host_info: HostInfo,
+    ) -> Result<Self> {
+        let display_name = DisplayName::new(name).map_err(|e| {
+            error!("Invalid DisplayName {:?}", e);
+            CommonError::FactorSourceNameInvalid
+        })?;
+        let id_from_hash =
+            FactorSourceIDFromHash::from_mnemonic_with_passphrase(
+                factor_source_kind,
+                &mnemonic_with_passphrase,
+            );
+
+        match factor_source_kind {
+            FactorSourceKind::Device => {
+                let is_main = false;
+                let hint = DeviceFactorSourceHint::with_info_and_label(
+                    &host_info,
+                    mnemonic_with_passphrase.mnemonic.word_count,
+                    display_name.value(),
+                );
+                Ok(FactorSource::from(DeviceFactorSource::babylon_with_hint(
+                    is_main,
+                    id_from_hash,
+                    hint,
+                )))
+            }
+            FactorSourceKind::OffDeviceMnemonic => {
+                let hint = OffDeviceMnemonicHint::new(
+                    display_name,
+                    mnemonic_with_passphrase.mnemonic.word_count,
+                );
+                Ok(FactorSource::from(OffDeviceMnemonicFactorSource::new(
+                    id_from_hash,
+                    hint,
+                )))
+            }
+            FactorSourceKind::Password => {
+                let hint = PasswordFactorSourceHint::new(display_name.value());
+                Ok(FactorSource::from(PasswordFactorSource::new(
+                    id_from_hash,
+                    hint,
+                )))
+            }
+            _ => Err(CommonError::InvalidFactorSourceKind {
```

Comment:
I think best avoid `_` and spell out the variants with `|` i.e. `FactorSourceKind::Arculus | FactorSourceKind::Ledger ` etc so that we get a compilation error here when we add a new kind, reminding us to handle that new variant.

Reply:
Yes we have used `_` in the past for `match` but I think many of those places ought to be changed to.